firedrake/projection.py

```diff
@@ -48,32 +50,56 @@ def check_meshes(source, target):
 
 @PETSc.Log.EventDecorator()
 @annotate_project
-def project(v, V, bcs=None,
-            solver_parameters=None,
-            form_compiler_parameters=None,
-            use_slate_for_inverse=True,
-            name=None,
-            ad_block_tag=None):
-    """Project a UFL expression into a :class:`.FunctionSpace`
+def project(
+    v: ufl.core.expr.Expr,
+    V: Union[firedrake.functionspaceimpl.FunctionSpace, firedrake.Function],
```

Comment:
```suggestion
    V: Union[firedrake.functionspaceimpl.WithGeometry, firedrake.Function],
```
`firedrake.functionspaceimpl.FunctionSpace` is topological.